ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs

```diff
@@ -170,17 +194,64 @@ instance Show (GenTxId (ShelleyBlock era)) where
   Applying transactions
 -------------------------------------------------------------------------------}
 
-applyShelleyTx ::
+applyShelleyTx :: forall era.
      ShelleyBasedEra era
   => LedgerConfig (ShelleyBlock era)
   -> SlotNo
   -> GenTx (ShelleyBlock era)
   -> TickedLedgerState (ShelleyBlock era)
+  -> Except (ApplyTxErr (ShelleyBlock era))
+       ( TickedLedgerState (ShelleyBlock era)
+       , Validated (GenTx (ShelleyBlock era))
+       )
+applyShelleyTx cfg slot (ShelleyTx _ tx) st = do
+    (mempoolState', vtx) <-
+       SL.applyTx
+         (shelleyLedgerGlobals cfg)
+         (SL.mkMempoolEnv   innerSt slot)
+         (SL.mkMempoolState innerSt)
+         tx
+
+    let st' = set theLedgerLens mempoolState' st
+
+    pure (st', mkShelleyValidatedTx vtx)
+  where
+    innerSt = tickedShelleyLedgerState st
+
+applyShelleyValidatedTx :: forall era.
+     ShelleyBasedEra era
+  => LedgerConfig (ShelleyBlock era)
+  -> SlotNo
+  -> Validated (GenTx (ShelleyBlock era))
+  -> TickedLedgerState (ShelleyBlock era)
   -> Except (ApplyTxErr (ShelleyBlock era)) (TickedLedgerState (ShelleyBlock era))
-applyShelleyTx cfg slot (ShelleyTx _ tx) st =
-    (\state -> st { tickedShelleyLedgerState = state }) <$>
-        SL.applyTxs
-          (shelleyLedgerGlobals cfg)
-          slot
-          (Seq.singleton tx)
-          (tickedShelleyLedgerState st)
+applyShelleyValidatedTx cfg slot (ShelleyValidatedTx _ tx) st = do
+    mempoolState' <-
+       SL.applyTxInBlock
+         (shelleyLedgerGlobals cfg)
+         (SL.mkMempoolEnv   innerSt slot)
+         (SL.mkMempoolState innerSt)
+         tx
+
+    pure $ set theLedgerLens mempoolState' st
+  where
+    innerSt = tickedShelleyLedgerState st
+
+-- | The lens combinator
+set ::
```

Comment:
This is tiny, but I'd rather import it. But I'm not convinced it justifies extending `build-depends` in the `.cabal` file. Hence what's here.